### DISCUSSION TITLE
Help with channeling constraints

### DISCUSSION THREAD

**Discussion by aihiangng** [google/or-tools#3268] [CP-SAT questions]
Not sure if I am using channeling constraints right, but I cant seem to get the right solution. 

I have the following requirements:

If previous month balance >=50, and previous month state is 1 then I want to set current month state can be 1
Months can have states: 1, 2, 3

```py
month_balance = dict()
for mth in range(100):
    month_balance[mth] = model.NewIntVar(-100, 5000, 'month_%i_balance' % (mth))
    
month_states = dict()
states_lst = [1, 2, 3]
for mth in range(100):
    for states in states_lst:
        month_states[mth, states] = model.NewBoolVar('month_%i_state%i' % (mth, states))


for curr_mth in range(1, 100):
    c_temp[curr_mth] = model.NewBoolVar('c_month%i' % (curr_mth)) # I created a temp variable to help me channel my constraint. 
    model.Add(month_balance[curr_mth-1] >= 50).OnlyEnforceIf(c_temp[curr_mth]) # curr_mth-1 to check previous month
    model.Add(month_balance[curr_mth-1]< 50).OnlyEnforceIf(c_temp[curr_mth].Not())
    # If previous month state is 1 and value more than 50, then current month set month to 1
    # Does the line below convey this idea?
    model.Add(month_states[curr_mth, 1]==1).OnlyEnforceIf([month_states[curr_mth-1, 1], c_temp[curr_mth]])
    # I am also assuming that whatever boolean I put in OnlyEnforceIf([bool1, bool2, ...]), it will make sure all is 1 with AND condition
```

**Comment by prissky** [comment]
Shouldn't the last line be replaced (complemented) by a second temp variable representing that the month_states[curr_mth - 1] == 1?

**Reply by aihiangng** [reply]
@prissky Hmmwhy would I need that if my month_states takes on boolean? 


month_states[curr_mth-1, 1] > if this is true, its 1 AND c_temp[curr_mth] is true, then => month_states[curr_mth, 1]==1

**Reply by prissky** [reply]
IINM, to ensure that each month takes on one state. Does the channeling provide that in your model? Consider a situation when month_balance[1] < 50 for example.

**Reply by aihiangng** [reply]
@prissky Hmm in the situation when month_balance[1] < 50, then current month states can be 1, 2, or 3 doesnt matter. But only when previous month balance > 50 and previous month state=1, curr month state must be 1

**Reply by lperron** [reply]
don't you need an AddExactlyOne(month_states[mth, states] for all states) ?

**Reply by lperron** [reply]
```
model.Add(month_states[curr_mth, 1]==1).OnlyEnforceIf([month_states[curr_mth-1, 1], c_temp[curr_mth]])
```

you should stay in the Boolean world.

```
model.AddBoolOr(month_states[curr_mth, 1]).OnlyEnforceIf([month_states[curr_mth-1, 1], c_temp[curr_mth]])
```

**Reply by aihiangng** [reply]
@lperron does using AddBoolOr speed things up or use less memory?

**Comment by aihiangng** [comment]
@lperron Hey, do you mind giving advice on how the channeling for the above constraint can be implemented? Thank you ((:

**Comment by lperron** [comment]
A bit of both at the modeling level. After presolve, it will be the same.
